### PR TITLE
Introduce interface for relay chain interaction

### PR DESCRIPTION
**Pull request by skunert**
In this PR, I introduce an interface that exposes methods for interaction with the relay chain, as discussed in #545.
These changes are only refactoring and have no impact on functionality.
Outline of the changes:
- Introduce trait `RelayChainInterface`
- Implement `RelayChainInterface` for `RelayChainLocal`, a status-quo implementation that uses a full node in the same process
- replace all occurrences of polkadot client and backend with `RelayChainLocal`

### CODE REVIEW

File: client/collator/src/lib.rs [hunk 1]
```diff
@@ -34,9 +36,10 @@ use polkadot_node_primitives::{
 };
 use polkadot_node_subsystem::messages::{CollationGenerationMessage, CollatorProtocolMessage};
 use polkadot_overseer::Handle as OverseerHandle;
-use polkadot_primitives::v1::{CollatorPair, Hash as PHash, HeadData, Id as ParaId};
+use polkadot_primitives::v1::{CollatorPair, HeadData, Id as ParaId};
 
 use codec::{Decode, Encode};
+use cumulus_primitives_core::CollectCollationInfo;
```
Comment (bkchr) [inline client/collator/src/lib.rs:42]: Why was this moved down?

File: client/consensus/aura/src/lib.rs [hunk 2]
```diff
@@ -46,11 +45,12 @@ use sp_consensus_aura::AuraApi;
...
 use sp_inherents::{CreateInherentDataProviders, InherentData, InherentDataProvider};
 use sp_keystore::SyncCryptoStorePtr;
-use sp_runtime::traits::{Block as BlockT, HashFor, Header as HeaderT, Member, NumberFor};
-use std::{convert::TryFrom, hash::Hash, marker::PhantomData, sync::Arc};
+use sp_runtime::traits::{Block as BlockT, Header as HeaderT, Member, NumberFor};
+use std::{convert::TryFrom, hash::Hash, sync::Arc};
 
 mod import_queue;
 
+use cumulus_relay_chain_interface::RelayChainInterface;
```
Comment (bkchr) [inline client/consensus/aura/src/lib.rs:53]: Please move to the top to the other imports.

File: client/consensus/aura/Cargo.toml [hunk 3]
```diff
@@ -24,12 +24,11 @@ sc-telemetry = { git = "https://github.com/paritytech/substrate", branch = "mast
 sc-consensus = { git = "https://github.com/paritytech/substrate", branch = "master" }
 substrate-prometheus-endpoint = { git = "https://github.com/paritytech/substrate", branch = "master" }
 
-# Polkadot dependencies
-polkadot-client = { git = "https://github.com/paritytech/polkadot", branch = "master" }
-
 # Cumulus dependencies
 cumulus-client-consensus-common = { path = "../common" }
+cumulus-client-collator = { path = "../../../client/collator" }
```
Comment (bkchr) [inline client/consensus/aura/Cargo.toml:29]: Why do we need this?

Comment (bkchr) [reply client/consensus/aura/Cargo.toml:29]: Looks unused to me.

File: client/consensus/common/Cargo.toml [hunk 4]
```diff
@@ -18,6 +18,8 @@ sp-trie = { git = "https://github.com/paritytech/substrate", branch = "master" }
 # Polkadot deps
 polkadot-primitives = { git = "https://github.com/paritytech/polkadot", branch = "master" }
 
+cumulus-relay-chain-interface = { path = "../../relay-chain-interface" }
```
Comment (bkchr) [inline client/consensus/common/Cargo.toml:21]:
```suggestion
# Cumulus deps
cumulus-relay-chain-interface = { path = "../../relay-chain-interface" }
```

File: client/relay-chain-interface/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "cumulus-relay-chain-interface"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (bkchr) [inline client/relay-chain-interface/Cargo.toml:6]:
```suggestion
```

File: client/relay-chain-interface/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "cumulus-relay-chain-interface"
+version = "0.1.0"
+edition = "2021"
```
Comment (bkchr) [inline client/relay-chain-interface/Cargo.toml:4]: Author missing. Please check the other Cargo.tomls on what we set by default.

File: client/pov-recovery/src/lib.rs [hunk 7]
```diff
@@ -420,19 +418,17 @@ where
...
 /// Returns a stream over pending candidates for the parachain corresponding to `para_id`.
 fn pending_candidates<RC>(
-	relay_chain_client: Arc<RC>,
+	relay_chain_client: RC,
 	para_id: ParaId,
 ) -> impl Stream<Item = (CommittedCandidateReceipt, SessionIndex)>
 where
-	RC: ProvideRuntimeApi<PBlock> + BlockchainEvents<PBlock>,
-	RC::Api: ParachainHost<PBlock>,
+	RC: RelayChainInterface<PBlock> + Clone,
```
Comment (bkchr) [inline client/pov-recovery/src/lib.rs:425]: Do we need clone here?

File: client/pov-recovery/src/lib.rs [hunk 8]
```diff
@@ -420,19 +418,17 @@ where
 
 /// Returns a stream over pending candidates for the parachain corresponding to `para_id`.
 fn pending_candidates<RC>(
-	relay_chain_client: Arc<RC>,
+	relay_chain_client: RC,
```
Comment (bkchr) [inline client/pov-recovery/src/lib.rs:421]:
```suggestion
	relay_chain_client: impl RelayChainInterface,
```

If not we can directly do this.

File: client/network/src/wait_on_relay_chain_block.rs [hunk 9]
```diff
@@ -121,40 +122,66 @@ where
...
+
+impl SyncOracle for DummyNetwork {
+	fn is_major_syncing(&mut self) -> bool {
+		unimplemented!("Not needed for test")
+	}
+
+	fn is_offline(&mut self) -> bool {
+		unimplemented!("Not needed for test")
+	}
+}
```
Comment (bkchr) [inline client/network/src/wait_on_relay_chain_block.rs:135]: Should be moved into the test module.

File: client/network/src/lib.rs [hunk 10]
```diff
@@ -408,87 +384,37 @@ where
...
 
 /// Block announce validator builder.
 ///
-/// Builds a [`BlockAnnounceValidator`] for a parachain. As this requires
-/// a concrete relay chain client instance, the builder takes a [`polkadot_client::Client`]
-/// that wraps this concrete instanace. By using [`polkadot_client::ExecuteWithClient`]
-/// the builder gets access to this concrete instance.
-struct BlockAnnounceValidatorBuilder<Block, B> {
+/// Builds a [`BlockAnnounceValidator`] for a parachain.
+struct BlockAnnounceValidatorBuilder<Block, RCInterface> {
```
Comment (bkchr) [inline client/network/src/lib.rs:400]: We don't need the builder anymore. The function above should be enough.

File: client/network/src/lib.rs [hunk 11]
```diff
@@ -230,52 +225,39 @@ impl TryFrom<&'_ CollationSecondedSignal> for BlockAnnounceData {
 /// chain. If it is at the tip, it is required to provide a justification or otherwise we reject
 /// it. However, if the announcement is for a block below the tip the announcement is accepted
 /// as it probably comes from a node that is currently syncing the chain.
-pub struct BlockAnnounceValidator<Block, R, B, BCE> {
+pub struct BlockAnnounceValidator<Block, RCInterface> {
```
Comment (bkchr) [inline client/network/src/lib.rs:228]: Sometimes you call it `R`, sometimes `RCInterface`. Maybe a little bit consistency would be nice :P

File: client/relay-chain-interface/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+	let (full_node, collator_key) =
+		build_polkadot_full_node(polkadot_config, telemetry_worker_handle).map_err(
+			|e| match e {
+				polkadot_service::Error::Sub(x) => x,
+				s => format!("{}", s).into(),
+			},
+		)?;
+
+	let sync_oracle: Box<dyn SyncOracle + Send + Sync> = Box::new(full_node.network.clone());
+	let network = Arc::new(Mutex::new(sync_oracle));
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:431]: Sync oracle changes its name to network?

We should probably just call it `sync_oracle`

File: client/relay-chain-interface/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+	fn overseer_handle(&self) -> Option<Handle>;
+}
+
+/// RelayChainLocal is used to interact with a full node that is running locally
+/// in the same process.
+pub struct RelayChainLocal<Client> {
+	pub full_client: Arc<Client>,
+	pub backend: Arc<FullBackend>,
+	pub network: Arc<Mutex<Box<dyn SyncOracle + Send + Sync>>>,
+	pub overseer_handle: Option<Handle>,
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:102]: Why are all of them public?

Would be better to provide a `new` function.

File: client/relay-chain-interface/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+use sc_telemetry::TelemetryWorkerHandle;
+use sp_api::{ApiError, BlockT, ProvideRuntimeApi};
+use sp_consensus::SyncOracle;
+use sp_core::{sp_std::collections::btree_map::BTreeMap, Pair};
+use std::sync::Mutex;
+
+const LOG_TARGET: &str = "relay-chain-interface";
+
+/// Should be used for all interaction with the relay chain in cumulus.
+pub trait RelayChainInterface<Block: BlockT> {
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:30]: Why is that generic over the block?

Comment (bkchr) [reply client/relay-chain-interface/src/lib.rs:30]: And we could here directly require `Send + Sync` as super traits. Then we don't need to add this when we want to use this trait.

File: client/relay-chain-interface/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,441 @@
+use std::{sync::Arc, time::Duration};
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:1]: Missing license header.

**Review by bkchr — COMMENTED** [review]
Have given it a first review pass. Looks already really good, just some stuff here and there that should be improved.

File: client/relay-chain-local/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+use sc_telemetry::TelemetryWorkerHandle;
+use sp_api::{ApiError, ProvideRuntimeApi};
+use sp_consensus::SyncOracle;
+use sp_core::{sp_std::collections::btree_map::BTreeMap, Pair};
+use sp_state_machine::{Backend as StateBackend, StorageValue};
+use std::sync::Mutex;
+const LOG_TARGET: &str = "relay-chain-local";
+
+/// RelayChainLocal is used to interact with a full node that is running locally
+/// in the same process.
```
Comment (bkchr) [inline client/relay-chain-local/src/lib.rs:46]:
```suggestion
/// Provides an implementation of the [`RelayChainInterface`] using a local in-process relay chain node.
```

File: client/relay-chain-local/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+				)
+			})
+			.ok()?;
+		state
+			.storage(key)
+			.map_err(|e| {
+				tracing::error!(
+					target: LOG_TARGET,
+					error = ?e,
+					"Cannot decode the hrmp ingress channel index.",
```
Comment (bkchr) [inline client/relay-chain-local/src/lib.rs:216]: C&P? :P

File: client/relay-chain-local/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+				.map_err(|e| {
+					tracing::error!(
+						target: LOG_TARGET,
+						relay_parent = ?block_id,
+						error = ?e,
+						"Failed to collect required relay chain state storage proof.",
+					);
+					e
+				})
+				.map(|v| Some(v)),
```
Comment (bkchr) [inline client/relay-chain-local/src/lib.rs:251]:
```suggestion
				.map(Some),
```

File: client/relay-chain-local/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+			_ => {},
+		}
+
+		let listener = self.full_client.import_notification_stream();
+
+		// Now it is safe to drop the lock, even when the block is now imported, it should show
+		// up in our registered listener.
+		drop(_lock);
+
+		Ok(BlockCheckResult::NotFound(listener))
```
Comment (bkchr) [inline client/relay-chain-local/src/lib.rs:274]: Maybe we should instead directly return a future that will resolve when the block is imported?

File: client/relay-chain-local/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+	}
+}
+
+/// Builder for a concrete relay chain interface, creatd from a full node. Builds
+/// a [`RelayChainLocal`] to access relay chain data necessary for parachain operation.
+///
+/// The builder takes a [`polkadot_client::Client`]
+/// that wraps a concrete instance. By using [`polkadot_client::ExecuteWithClient`]
+/// the builder gets access to this concrete instance and instantiates a RelayChainLocal with it.
```
Comment (bkchr) [inline client/relay-chain-local/src/lib.rs:283]:
```suggestion
/// the builder gets access to this concrete instance and instantiates a [`RelayChainLocal`] with it.
```

File: parachain-template/node/src/service.rs [hunk 21]
```diff
@@ -18,8 +18,11 @@ use cumulus_client_service::{
 	prepare_node_config, start_collator, start_full_node, StartCollatorParams, StartFullNodeParams,
 };
 use cumulus_primitives_core::ParaId;
+use cumulus_relay_chain_interface::RelayChainInterface;
+use cumulus_relay_chain_local::build_relay_chain_interface;
 
 // Substrate Imports
+use core::time::Duration;
```
Comment (bkchr) [inline parachain-template/node/src/service.rs:25]: Please move this up to the std import and import it from std.

File: polkadot-parachains/src/service.rs [hunk 22]
```diff
@@ -27,11 +28,14 @@ use cumulus_primitives_core::{
...
 	ParaId,
 };
+use cumulus_relay_chain_interface::RelayChainInterface;
+use cumulus_relay_chain_local::build_relay_chain_interface;
 use polkadot_service::NativeExecutionDispatch;
 
 use crate::rpc;
 pub use parachains_common::{AccountId, Balance, Block, Hash, Header, Index as Nonce};
 
+use core::time::Duration;
```
Comment (bkchr) [inline polkadot-parachains/src/service.rs:38]: Same

File: polkadot-parachains/src/service.rs [hunk 23]
```diff
@@ -129,7 +133,9 @@ impl sc_executor::NativeExecutionDispatch for StatemineRuntimeExecutor {
 	}
 }
 
-/// Native Westmint executor instance.
+/**
+Native Westmint executor instance.
+*/
```
Comment (bkchr) [inline polkadot-parachains/src/service.rs:138]:
```suggestion
/// Native Westmint executor instance.
```

File: polkadot-parachains/Cargo.toml [hunk 24]
```diff
@@ -74,8 +74,11 @@ cumulus-client-consensus-relay-chain = { path = "../client/consensus/relay-chain
 cumulus-client-consensus-common = { path = "../client/consensus/common" }
 cumulus-client-service = { path = "../client/service" }
 cumulus-client-network = { path = "../client/network" }
+cumulus-client-collator = { path = "../client/collator" }
```
Comment (bkchr) [inline polkadot-parachains/Cargo.toml:77]: Do we need this?

File: test/service/src/lib.rs [hunk 25]
```diff
@@ -21,14 +21,17 @@
 mod chain_spec;
 mod genesis;
 
-use core::future::Future;
+use core::{future::Future, time::Duration};
```
Comment (bkchr) [inline test/service/src/lib.rs:24]:
```suggestion
use std::{future::Future, time::Duration};
```

File: test/service/Cargo.toml [hunk 26]
```diff
@@ -29,6 +29,7 @@ sc-transaction-pool = { git = "https://github.com/paritytech/substrate", branch
 sp-arithmetic = { git = "https://github.com/paritytech/substrate", branch = "master" }
 sp-blockchain = { git = "https://github.com/paritytech/substrate", branch = "master" }
 sp-core = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sp-api = { git = "https://github.com/paritytech/substrate", branch = "master" }
```
Comment (bkchr) [inline test/service/Cargo.toml:32]: We don't need this?

File: test/service/Cargo.toml [hunk 27]
```diff
@@ -47,11 +48,13 @@ polkadot-test-service = { git = "https://github.com/paritytech/polkadot", branch
 cumulus-client-consensus-relay-chain = { path = "../../client/consensus/relay-chain" }
 cumulus-client-network = { path = "../../client/network" }
 cumulus-client-service = { path = "../../client/service" }
+cumulus-client-collator = { path = "../../client/collator" }
```
Comment (bkchr) [inline test/service/Cargo.toml:51]: And this?

File: primitives/parachain-inherent/src/client_side.rs [hunk 28]
```diff
@@ -181,38 +89,26 @@ fn collect_relay_storage_proof(
...
-		.map_err(|e| {
-			tracing::error!(
-				target: LOG_TARGET,
-				relay_parent = ?relay_parent,
-				error = ?e,
-				"Failed to collect required relay chain state storage proof.",
-			)
-		})
-		.ok()
+	relay_chain_interface.prove_read(&relay_parent_block_id, &relevant_keys).ok()?
```
Comment (bkchr) [inline primitives/parachain-inherent/src/client_side.rs:92]: Weird that you get here a result and convert it to Option and below you just return Option directly.

File: client/relay-chain-interface/src/lib.rs [hunk 29]
```diff
@@ -0,0 +1,210 @@
...
+		v1::{CommittedCandidateReceipt, OccupiedCoreAssumption, SessionIndex, ValidatorId},
+		Block as PBlock, BlockId, Hash as PHash, InboundHrmpMessage,
+	},
+	InboundDownwardMessage, ParaId, PersistedValidationData,
+};
+use polkadot_overseer::Handle as OverseerHandle;
+use sc_client_api::{blockchain::BlockStatus, StorageProof};
+
+use sp_api::ApiError;
+use sp_core::sp_std::collections::btree_map::BTreeMap;
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:30]: `std::collections::BTreeMap`

Comment (bkchr) [reply client/relay-chain-interface/src/lib.rs:30]: Should be used here.

File: client/relay-chain-interface/src/lib.rs [hunk 30]
```diff
@@ -0,0 +1,210 @@
...
+use sp_state_machine::StorageValue;
+
+pub enum BlockCheckResult {
+	InChain,
+	NotFound(sc_client_api::ImportNotifications<PBlock>),
+}
+
+/// Should be used for all interaction with the relay chain in cumulus.
+pub trait RelayChainInterface: Send + Sync {
+	fn get_storage_by_key(&self, block_id: &BlockId, key: &[u8]) -> Option<StorageValue>;
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:40]: Docs?

Also weird that we return `None` here for when the key doesn't exist or there was an error.

File: client/relay-chain-interface/src/lib.rs [hunk 31]
```diff
@@ -0,0 +1,210 @@
...
+
+	fn validators(&self, block_id: &BlockId) -> Result<Vec<ValidatorId>, ApiError>;
+
+	fn block_status(&self, block_id: BlockId) -> Result<BlockStatus, sp_blockchain::Error>;
+
+	fn best_block_hash(&self) -> PHash;
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:46]: Docs please

File: client/relay-chain-interface/src/lib.rs [hunk 32]
```diff
@@ -0,0 +1,210 @@
...
+		para_id: ParaId,
+	) -> Result<Option<CommittedCandidateReceipt>, ApiError>;
+
+	fn session_index_for_child(&self, block_id: &BlockId) -> Result<SessionIndex, ApiError>;
+
+	fn import_notification_stream(&self) -> sc_client_api::ImportNotifications<PBlock>;
+
+	/// Check if block is in chain. If it is, we return BlockCheckResult::InChain.
+	/// If it is not in the chain, we return BlockCheckResult::NotFound with a listener that can be used to wait on the block.
+	fn check_block_in_chain(
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:87]: Part of the docs should go to the `BlockCheckResult` type.

However, I'm not really happy with the naming here.

Maybe something like `wait_for_block` and return a future that either resolves directly or waits for the block to be imported.

File: client/relay-chain-interface/Cargo.toml [hunk 33]
```diff
@@ -0,0 +1,20 @@
+[package]
+authors = ["Parity Technologies <admin@parity.io>"]
+name = "cumulus-relay-chain-interface"
+version = "0.1.0"
+edition = "2021"
+
```
Comment (bkchr) [inline client/relay-chain-interface/Cargo.toml:6]:
```suggestion
```

File: client/network/src/lib.rs (hunk 10 above)
Comment (bkchr) [reply client/network/src/lib.rs:401]: Still relevant.

File: primitives/parachain-inherent/src/client_side.rs (hunk 28 above)
Comment (skunert) [reply primitives/parachain-inherent/src/client_side.rs:110]: I made this consistent now, If an error is possible, I return `Result` from the interface

File: client/relay-chain-local/src/lib.rs [hunk 34]
```diff
@@ -0,0 +1,542 @@
...
+
+pub enum BlockCheckStatus {
+	/// Block is in chain
+	InChain,
+	/// Block status is unknown, listener can be used to wait for notification
+	Unknown(ImportNotifications<PBlock>),
+}
+
+// Helper function to check if a block is in chain.
+pub fn check_block_in_chain<Client>(
```
Comment (skunert) [inline client/relay-chain-local/src/lib.rs:295]: This helper function came back to solve an issue which the lock guard that was not `Send` in the async function. Also makes this reusable in the test.

File: primitives/parachain-inherent/src/client_side.rs [hunk 35]
```diff
@@ -19,112 +19,34 @@
...
+		.get_storage_by_key(
+			&relay_parent_block_id,
+			&relay_well_known_keys::hrmp_ingress_channel_index(para_id),
+		)
 		.map_err(|e| {
 			tracing::error!(
 				target: LOG_TARGET,
-				error = ?e,
-				"Cannot obtain the hrmp ingress channel index."
+				relay_parent = ?relay_parent_block_id,
```
Comment (bkchr) [inline primitives/parachain-inherent/src/client_side.rs:48]:
```suggestion
				relay_parent = ?relay_parent,
```

File: primitives/parachain-inherent/src/client_side.rs [hunk 36]
```diff
@@ -19,112 +19,34 @@
...
+				error = ?e, "Cannot obtain the hrmp ingress channel."
```
Comment (bkchr) [inline primitives/parachain-inherent/src/client_side.rs:49]:
```suggestion
				error = ?e,
				"Cannot obtain the hrmp ingress channel.",
```

File: client/network/src/lib.rs [hunk 37]
```diff
@@ -408,88 +379,14 @@ where
 /// Build a block announce validator instance.
 ///
 /// Returns a boxed [`BlockAnnounceValidator`].
-pub fn build_block_announce_validator<Block: BlockT, B>(
-	relay_chain_client: polkadot_client::Client,
+pub fn build_block_announce_validator<Block: BlockT, RCInterface>(
```
Comment (bkchr) [inline client/network/src/lib.rs:382]: Function is rather useless now, please remove.

File: client/consensus/aura/src/lib.rs [hunk 38]
```diff
@@ -74,44 +69,39 @@ pub struct AuraConsensus<B, RClient, RBackend, CIDP> {
...
+			backoff_authoring_blocks,
+			sync_oracle,
+			keystore,
+			force_authoring,
+			slot_duration,
+			telemetry,
+			block_proposal_slot_portion,
+			max_block_proposal_slot_portion,
+		}: BuildAuraConsensusParams<PF, BI, CIDP, Client, BS, SO>,
+	) -> Box<dyn ParachainConsensus<B>>
```
Comment (bkchr) [inline client/consensus/aura/src/lib.rs:104]:
```suggestion
	) -> Self
```

We don't need to box anymore. This was just a short coming of the old implementation with all the `ExecuteWithClient` stuff etc, that is now solved.

**Review by bkchr — APPROVED** [review]
Looks good! Just some last small nitpicks.
